code/bugfixes/fix_pe_fieldmaps.py

```diff
@@ -0,0 +1,127 @@
+# emacs: -*- mode: python; py-indent-offset: 4; indent-tabs-mode: nil -*-
+# vi: set ft=python sts=4 ts=4 sw=4 et:
+#
+# Copyright 2024 The Axon Lab <theaxonlab@gmail.com>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# We support and encourage derived works from this project, please read
+# about our expectations at
+#
+#     https://www.nipreps.org/community/licensing/
+"""Fix PA fieldmaps which are AP in reality."""
+from __future__ import annotations
+
+from pathlib import Path
+from json import loads
+import subprocess
+
+import pandas as pd
+from datalad.api import unlock
+
+
+def check_and_fix(json_path: str | Path) -> bool:
+    """
+    Check and fix wrong PA.
+
+    Parameters
+    ----------
+    json_path : :obj:`os.pathlike`
+        Path to the JSON file.
+
+    Returns
+    -------
+    bool
+        True if fix was applied, False otherwise.
+
+    """
+
+    json_path = Path(json_path)
+
+    metadata = loads(json_path.read_text())
+
+    if metadata["PhaseEncodingDirection"] != "j-":
+        return False
+
+    print(f"Fixing <{json_path}>")
+    nifti_path = str(json_path).replace(".json", ".nii.gz")
+    json_ap = str(json_path).replace("_dir-PA_epi", "_dir-AP_epi")
+    ap_exists = Path(json_ap).exists()
+
+    # Read scans_tsv
+    scans_tsv = (
+        json_path.parent.parent
+        / json_path.name.replace("_acq-b0_dir-PA_epi.json", "_scans.tsv")
+    )
+    scans_db = pd.read_csv(scans_tsv, sep=r"\s+", index_col=None)
+
+    run_entity = ""
+    if ap_exists:
+        # Sometimes, the "PA" direction was acquired after the AP
```

Comment:
It is not clear to me what you mean by "Sometimes, the "PA" direction was acquired after the AP" because I thought one session was supposed to acquire one phase-encoding direction.

Reply:
> It is not clear to me what you mean by "Sometimes, the "PA" direction was acquired after the AP" because I thought one session was supposed to acquire one phase-encoding direction.

Great you are asking - there are things we need to refresh. In the end, considering that acquiring fieldmaps is pretty cheap, we went ahead and collected the 4 possible PE directions for both the b0 EPI and the bold EPI. You can see that in the protocol printouts (links within the blue box here https://www.axonlab.org/hcph-sops/data-collection/intro/#scanning-protocols). We did this after session 4, if I remember correctly.

As @acionca noticed, if you go to the [HCPh_PA](https://www.axonlab.org/hcph-sops/assets/files/HCPh_PA.pdf) protocol, you will find that the b0 fieldmap for the PA direction is configured with phase encoding A >>> P (AP) while it should be P >>> A. That is the error.

> I understand that under `if ap_exists`, you assign run_1 and run_2, but I don't understand why the AP direction should already exists.

The problem here is that, in some sessions AP was acquired first, and others PA was first. Once we realized of my mistake with the PE direction, for the b0 fieldmaps all the `dir-PA` runs needed to be renamed to be `dir-AP` to reflect the actual direction. However, one `dir-AP` already exists. This `if ap_exists` is necessary to add the `run-1` or `run-2` entity. For piloting sessions it is possible that we acquired `dir-PA` and there is not a `dir-AP`, so in that case you should not add `run-`.

Once you have confirmed you do need to add `run-`, then you need to identify if the `dir-PA` was acquired first or the existing `dir-AP` was first. Since both will receive a new `run-` entity, you want to make sure that `run-1` is assigned to the first that was acquired and `run-2` to the second.

Does it make sense?

Reply:
Ok I was missing the information that you acquired the fieldmap with the 4 possible PE directions each session. Now it makes sense. Thanks